### PR TITLE
Introduce logging middleware

### LINKED ISSUE #4941: Rework logger as a driver middleware
The current `SQLLogger` has major disadvantages:
1. It doesn't log exceptions (https://github.com/doctrine/dbal/issues/3228) or anything else (https://github.com/doctrine/dbal/pull/3513) besides what's hard-coded in the wrapper connection.
2. It's not extensible. Any addition to the interface would be a breaking change.
3. It's not PSR-compliant.
4. It's not a logger (https://github.com/doctrine/dbal/pull/3513#issuecomment-483062136).

As of DBAL 3.0 (https://github.com/doctrine/dbal/pull/4157), the `Middleware` interface is available which could be used to implement logging. It could log connecting, disconnecting, executing queries and prepared statements, numbers of fetched and affected rows, errors, and all other details that are available at the driver level API. The end-users will be able to extend it by implementing their own logging functionality without the need to change the API.

### PR DESCRIPTION
**Pull request by morozov**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no

Closes #4941.

This middleware is meant to replace the `SQLLogger` interface but it's not a drop-in replacement:
1. It will log the statements and their parameters in the form in which they are passed to the DBAL driver (i.e. with list parameters expanded and named parameters converted to the positional ones).
2. There is no equivalent of the `stopQuery()` method. If instrumentation around the query methods is needed, it should be implemented as a separate middleware.

**Technical details**:
1. "Connecting" is logged in the driver but "Disconnecting" is logged in the connection. This is a bit asymmetric but there's a reason: the connection parameters in the DBAL format are only available at the driver level but the driver doesn't handle disconnection.
2. The middleware doesn't log exceptions. The severity of the exception as well as the need to log it should be determined by the component that handles the exception.

**Test changes**:
1. The IBM DB2 connection test is removed as it assumes that `Connection::getWrappedConnection()` will return a driver connection while it may return a middleware (see https://github.com/doctrine/dbal/pull/4966). The error handling during `prepare()` cannot always be tested in isolation and is covered in a more reliable way in https://github.com/doctrine/dbal/blob/07453f22fe1372c99d766f7ea591b7629d18ff76/tests/Functional/ConnectionTest.php#L413
2. The assertion in `testDeterminesDatabasePlatformWhenConnectingToNonExistentDatabase()` is removed since the middleware driver always implements the `VersionAwarePlatformDriver`, even if the wrapped driver doesn't, so based on that, the wrapper connection will always connect to detect the server version. This will be likely addressed by https://github.com/doctrine/dbal/pull/4764 in 4.0.0. UPD: this entire feature is pointless (https://github.com/doctrine/dbal/pull/4764#discussion_r697695742).

**TODO**:
- [x] Add tests for the logger middleware.
- [x] Remove the dependency of `ResultCacheTest` on `DebugStack` (https://github.com/doctrine/dbal/pull/5008).
- [X] Remove the "Temporarily enable Monolog in the test suite" commit.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -36,11 +36,13 @@
         "doctrine/cache": "^1.11|^2.0",
         "doctrine/deprecations": "^0.5.3",
         "doctrine/event-manager": "^1.0",
-        "psr/cache": "^1|^2|^3"
+        "psr/cache": "^1|^2|^3",
+        "psr/log": "^1|^2"
```
Comment (derrabus) [inline composer.json]:
```suggestion
        "psr/log": "^1|^2|^3"
```

Comment (morozov) [reply composer.json]: For some reason, allowing `psr/log: ^3` causes the downgrade of `symfony/console` to `v5.3.2`:
```
$ composer why-not symfony/console 5.3.10
symfony/console  v5.3.10  conflicts  psr/log (>=3)
```
In turn, this causes a [build failure](https://github.com/doctrine/dbal/runs/4205444243) on PHP 8.1 due to the issue fixed in https://github.com/symfony/console/commit/9a4fcecbcccf4b55f4c28e1c71e3cd83d7be99b8:
> Return type of Symfony\Component\Console\Helper\HelperSet::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice

~~This is weird because `symfony/console` requires `psr/log: ^1|^2` only as a development dependency.~~ UPD: it does explicitly conflict with `psr/log: >= 3` (https://github.com/symfony/console/commit/a96e44ed3144d6c68757f8a392c92f7056d46df9).

Comment (morozov) [reply composer.json]: The idea to introduce a conflict of `symfony/console` with `psr/log: >= 3` seems brutal: Symfony Console requires PSR Log only as a development dependency (i.e. doesn't depend on it in production) but the conflict affects all packages that use the log in production.

Comment (derrabus) [reply composer.json]: > Symfony Console requires PSR Log only as a development dependency

The package bundles a logger implementation that is incompatible with psr/log 3. That implementation is meant for production use as well.

The fact that Composer downgrades symfony/console is of course unfortunate. This will be resolved as soon as Symfony 6.0 is stable, I guess. But if we merge this PR as is, DBAL will block the installation of psr/log 3 as well. Somehow, we need to break out of this vicious circle. 🤔

Comment (morozov) [reply composer.json]: > The package bundles a logger implementation that is incompatible with psr/log 3. That implementation is meant for production use as well.

Is the logger bundled by Symfony named `\Psr\Log`? Could you provide some more details?

I believe, for testing purposes, we can artificially downgrade `psr/log` to 2 when testing on PHP 8.1 to allow the installation of the newest stable Symfony Console. It will not impact the production scenarios unless the production code depends on `symfony/console`. In this case, the developer will have to make the call about how to resolve the dependency constraints.

Comment (derrabus) [reply composer.json]: > Is the logger bundled by Symfony named `\Psr\Log`? Could you provide some more details?

The class is `Symfony\Component\Console\Logger\ConsoleLogger`. It extends `AbstractLogger` from the PSR package, but does not declare the return types that v3 requires.

> I believe, for testing purposes, we can artificially downgrade `psr/log` to 2 when testing on PHP 8.1 to allow the installation of the newest stable Symfony Console.

Let's do this then.

**Review by derrabus — DISMISSED** [review]

File: .github/workflows/continuous-integration.yml [hunk 2]
```diff
@@ -38,6 +38,10 @@ jobs:
         with:
           fetch-depth: 2
 
+      - name: "Temporarily remove support for PSR Log 3 on PHP 8.1"
+        run: 'sed -i "s/\"psr\/log\": \"^1|^2|^3\"/\"psr\/log\": \"^1|^2\"/" composer.json'
+        if: "${{ matrix.php-version=='8.1' }}"
```
Comment (greg0ire) [inline .github/workflows/continuous-integration.yml:43]: Why is this necessary? There is no occurrence of `psr/log` in `composer.json`.

Also, is it not something that could be more robust and legible with `composer require --no-update "psr/log ^1|^2"`?

File: src/Logging/Driver.php [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    /**
+     * @param array<string,mixed> $params Connection parameters
+     *
+     * @return array<string,mixed>
+     */
+    private function maskPassword(array $params): array
+    {
+        if (isset($params['password'])) {
+            $params['password'] = '******';
```
Comment (greg0ire) [inline src/Logging/Driver.php]: One might think we are giving an indication about the number of characters. How about this instead:
```suggestion
            $params['password'] = '<redacted>';
```

Comment (morozov) [reply src/Logging/Driver.php]: I discussed this with my colleague and we came to the same conclusion. Let's change it.

File: .github/workflows/continuous-integration.yml (hunk 2 above)
Comment (morozov) [reply .github/workflows/continuous-integration.yml:43]: > Why is this necessary?

See the discussion starting at https://github.com/doctrine/dbal/pull/4967#discussion_r748911418.

> There is no occurrence of `psr/log` in `composer.json`.

We require `psr/log` as of this patch.

> Also, is it not something that could be more robust and legible with `composer require --no-update "psr/log ^1|^2"`?

As far as I can tell, `ramsey/composer-install` doesn't support `require`, and we don't install Composer as a binary.

**Review by greg0ire — APPROVED** [review]

**Comment by derrabus** [comment]
Follow-up: doctrine/DoctrineBundle#1429

**Comment by Igor100** [comment]
Hello @morozov 
How to use this midleware instead the construction:
      _$doctrine = $this->getDoctrine();
      $doctrineConnection = $doctrine->getConnection();
      $stack = new \Doctrine\DBAL\Logging\DebugStack();
      $doctrineConnection->getConfiguration()->setSQLLogger($stack);
      ...... src sql using .......
      dd($stack)_

**Comment by morozov** [comment]
@Igor100 here's an example:
https://github.com/doctrine/dbal/blob/69a38c2970ac13c5d524703bc0e6732f10ad19f3/tests/TestUtil.php#L66-L71

**Comment by Tobion** [comment]
Hello. What's the recommended way of disabling sql logging for certain parts of the app?
Removing the sql logger from the connection configuration as done in https://github.com/doctrine/DoctrineBundle/issues/787
does not seem to work anymore. The middleware will always log.
See also https://github.com/symfony/symfony/issues/45974

**Comment by morozov** [comment]
If you're using Monolog, you can try doing `popHandler()` / `pushHandler()` in those parts that need logging to be disabled. Otherwise, it depends on the logger.

**Comment by mkopinsky** [comment]
If SQLLogger is deprecated in 3.2, does that mean that it's safe to continue using in `^3.0`?

**Comment by derrabus** [comment]
> If SQLLogger is deprecated in 3.2, does that mean that it's safe to continue using in `^3.0`?

Yes. It means your code will break when upgrading to 4.0.

**Comment by raziel057** [comment]
From user perspective, deprecating the SQLLogger seems a bit weird to me as we are not able anymore to change it after the connection is created according to different contexts. Maybe you could provide a way to change or remove the logger in the connection decorator created by the middleware?

**Comment by morozov** [comment]
The middleware is just an adapter between the DBAL driver and a PSR-3 logger. If the logger needs to implement any application-specific logic, it should be implemented in the logger. See https://github.com/doctrine/dbal/pull/4967#issuecomment-1092323700 for example.

**Comment by nayodahl** [comment]
Hello @morozov ,
How do you use the equivalent of the deprecated DebugStack(), if there is any ?
I'm ok with adding the Logging\Middleware to the connection's configuration like this : 

```php
$logger = new Logger('PHPUnit');
$logger->pushHandler(new StreamHandler('/tmp/dbal.log', Logger::DEBUG));
$logger->pushProcessor(new PsrLogMessageProcessor(null, true));

$configuration = (new Configuration())->setMiddlewares([new Logging\Middleware($logger)]);
$conn          = DriverManager::getConnection(self::getConnectionParams(), $configuration);
```

But then, how do you get the debugStack like the profiler does for exemple ?
I used to simply inject it in an eventListener and do stuff of the queries...

**Comment by morozov** [comment]
@nayodahl there's no equivalent of `DebugStack`. You can implement your own middleware that will wrap calls to the methods you're interested in in the code that captures the time their execution takes. Then you'll need to register both these middlewares. Probably the profiling one first (since you don't want to profile logging) and then the logging one.
